syntax = "proto2";

message ImportantWords {
    required string word = 1;
    required float importance = 2;
    repeated string wordalt = 3;
    // different case variants not including word, if any (importance is of all taken as a group)
}

message Label {
    // favored class is argmax(logits)
    // for sentiment: confidences in negative, positive, neutral. the one that is higher than others = the classification. for empty input string, logits is empty
    repeated float logits = 1;
    // if explain option is enabled, these will be provided as being words most decisive in choosing the favored class:
    repeated ImportantWords words = 2;
}

message LabelOptions {
    // TODO: unused
    // comment is default value
    optional bool explain = 1; // true
    optional float explain_epsilon = 2; // 0.05
    optional int32 explain_maxwords = 3; // 10
    optional float explain_maxwords_portion = 4; // 0.2
    optional bool explain_stopwords = 5; // false // include stopwords as possibly important
    optional bool explain_punctuation = 6; // false // include punctuation as possibly important
}

message Document {
    // the request
    required string document_id = 1;
    repeated string segments = 2; // sentences or lines without any terminating newline
    optional LabelOptions options = 3; // TODO: unused
}

message LabeledDocument {
    // the response
    required string document_id = 1;
    repeated Label labels = 2; // one to one with segments in Document
}

message LabelDocumentRequest {
    // TODO: unused, unnecessary? just send multiple Document and use unique document_id?
    required string request_id = 1;
    repeated Document document = 2;
    optional LabelOptions options = 3;
}
